File: backend/timeline-core/src/main/java/com/timeline/core/domain/Timeline.java
package com.timeline.core.domain;

import java.time.LocalDateTime;
import java.util.List;

public record Timeline(
        Long id,
        String title,
        String description,
        List<Category> categories,
        long eventYear,
        PrecisionLevel precisionLevel,
        Integer eventMonth,
        Integer eventDay,
        int sortOrder,
        LocalDateTime eventLocalDateTime,
        LocalDateTime eventUtcDateTime,
        String timeZone,
        Long uncertaintyYears,
        String location,
        String source,
        LocalDateTime createdAt,
        LocalDateTime updatedAt,
        String createdBy,
        String updatedBy,
        EventType eventType,
        Long endYear,
        Integer endMonth,
        Integer endDay
) {
    public Timeline {
        if (title == null || title.isBlank()) {
            throw new IllegalArgumentException("title must not be blank");
        }
        if (precisionLevel == null) {
            throw new IllegalArgumentException("precisionLevel must not be null");
        }
        if (eventType == null) {
            eventType = EventType.POINT;
        }
        if (categories == null) {
            categories = List.of();
        }
    }
}

File: frontend/src/types/timeline.ts
export type PrecisionLevel =
  | 'BILLION_YEARS'
  | 'HUNDRED_MILLION_YEARS'
  | 'TEN_MILLION_YEARS'
  | 'MILLION_YEARS'
  | 'HUNDRED_THOUSAND_YEARS'
  | 'TEN_THOUSAND_YEARS'
  | 'MILLENNIUM'
  | 'CENTURY'
  | 'DECADE'
  | 'YEAR'
  | 'MONTH'
  | 'DAY'
  | 'HOUR'
  | 'MINUTE'
  | 'SECOND';

export interface TimelineEvent {
  id: number;
  title: string;
  description: string;
  categoryIds: number[];
  categoryNames: string[];
  eventYear: number;
  precisionLevel: PrecisionLevel;
  eventMonth: number | null;
  eventDay: number | null;
  eventType: 'POINT' | 'RANGE';
  endYear?: number;
  endMonth?: number;
  endDay?: number;
  sortOrder: number;
  eventLocalDateTime: string | null;
  eventUtcDateTime: string | null;
  timeZone: string | null;
  uncertaintyYears: number | null;
  location: string | null;
  source: string | null;
  createdAt: string;
  updatedAt: string;
  createdBy: string;
  updatedBy: string;
}

export interface TimelineRequest {
  title: string;
  description?: string;
  categoryIds: number[];
  eventYear: number;
  precisionLevel: PrecisionLevel;
  eventMonth?: number;
  eventDay?: number;
  eventType?: 'POINT' | 'RANGE';
  endYear?: number;
  endMonth?: number;
  endDay?: number;
  sortOrder?: number;
  eventLocalDateTime?: string;
  eventUtcDateTime?: string;
  timeZone?: string;
  uncertaintyYears?: number;
  location?: string;
  source?: string;
}

export interface Category {
  id: number;
  name: string;
  description: string;
  createdAt: string;
  updatedAt: string;
  createdBy: string;
  updatedBy: string;
}

export interface SearchParams {
  fromYear?: number;
  toYear?: number;
  categoryId?: number;
  precisionLevel?: PrecisionLevel;
}
